src/ShopBundle/Basket/BasketVariableReplacer.php

```diff
@@ -0,0 +1,140 @@
+{
+    public const BASKET_HIDDEN_FIELDS_PLACEHOLDER = 'BASKETHIDDENFIELDS';
+    private const HIDDEN_FIELDS_SNIPPET = '@ChameleonSystemShop/snippets/ShopBundle/BasketForm/hiddenBasketFields.html.twig';
+
+    /**
+     * @var Environment
+     */
+    private $twigEnvironment;
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    public function __construct(
+        Environment $twigEnvironment,
+        LoggerInterface $logger)
+    {
+        $this->twigEnvironment = $twigEnvironment;
+        $this->logger = $logger;
+    }
+
+    /**
+     * handleRequest will be invoked on kernel.request and add the hidden fields to the replacer.
+     * On error it will log to the request channel and move on. It will not halt execution.
+     *
+     * @param GetResponseEvent $event
+     */
+    public function handleRequest(GetResponseEvent $event): void
+    {
+        $request = $event->getRequest();
+        $queryParameters = $request->query->all();
+        $paramsToRender = $this->filterKeys($queryParameters, ['basket']);
+        $paramsToRender = $this->flattenQueryParameters($paramsToRender);
+
+        try {
+            $hiddenFieldsHtml = $this->twigEnvironment->render(
+                self::HIDDEN_FIELDS_SNIPPET,
+                ['values' => $paramsToRender]
+            );
+            \TTools::AddStaticPageVariables([self::BASKET_HIDDEN_FIELDS_PLACEHOLDER => $hiddenFieldsHtml]);
+        } catch(Error $error) {
+            $this->logger->error(
+                sprintf('Error rendering hidden fields for basket forms: %s', $error->getMessage()),
+                ['exception' => $error]
+            );
+        }
+    }
+
+    /**
+     * flattenQueryParameters will flatten the query to a form that can be rendered to an html form.
+     *
+     * Example:
+     * [
+     *   "foo" => ["bar" => "baz"]
+     * ]
+     * will become:
+     * [
+     *   "foo[bar]" => "baz"
+     * ]
+     *
+     * It will work recursively for deeper nested structures.
+     *
+     * @param array $queryParameters
+     * @return array
+     */
+    private function flattenQueryParameters(array $queryParameters): array
+    {
+        $params = [];
+        foreach ($queryParameters as $key => $value) {
+            if (is_string($value)) {
+                $params[$key] = $value;
+                continue;
+            }
+            if (is_array($value)) {
+                $subParameters = [];
+                foreach ($value as $subKey => $subValue) {
+                    $subParameters[sprintf('%s[%s]', $key, $subKey)] = $subValue;
+                }
+                $subRendered = $this->flattenQueryParameters($subParameters);
+                foreach ($subRendered as $subKeyRendered => $subValueRendered) {
+                    $params[$subKeyRendered] = $subValueRendered;
+                }
+            }
+        }
+
+        return $params;
+    }
+
+    /**
+     * filterKeys will filter all the keys given.
```

Comment:
What kind of a filtering?

Reply:
I thought `filter` would be clear enough as the default way name for methods that do what this method does are named "filter". It is in the same family as "map" and "reduce"

Reply:
But filtering something always has a purpose - I don't see that here in "filter keys".

Neither in the name the doc or the implementation.

Reply:
The method takes the values to filter as an argument. It won't make any assumptions on the purpose.